### ISSUE TITLE
Promises allow a hole in the typesystem

### ISSUE THREAD

**Issue by jsmaniac** [racket/typed-racket#159]
In the code below, we have `(ann some-code (U 'a #f))`, which typechecks, although `some-code` actually produces `'b`. No contracts catches this error at runtime either.

```scheme
#lang racket

;; delay/name is a macro, so we wrap it with a function so we can type it:
(module untyped racket
  (provide delay/name/thunk)
  (define (delay/name/thunk f) (delay/name (f))))

;; Now we require/typed our function
(module typed-delay/name/thunk typed/racket
  (require/typed (submod ".." untyped) [delay/name/thunk (∀ (T) (→ (→ T) (Promise T)))])
  (provide delay/name/thunk))

;; This module shows the bug:
(module bug typed/racket
  (require (submod ".." typed-delay/name/thunk))
  
  (define mutable : (U 'a 'b) 'a)
  
  (define ab-mut (delay/name/thunk (λ () (begin0 mutable
                                                 (set! mutable 'b))))) ;; BAD BAD BAD!
  
  (ann (if (eq? (force ab-mut) 'a) ;; Here, (eq? (force ab-mut) 'a)
           (force ab-mut) ;; But here, (eq? (force ab-mut) 'b)
           #f)
       (U 'a #f))) ;; This typechecks, and throws no warning, but prints 'b !!!

(require 'bug)
```

**Comment by samth** [comment]
Aha, the problem here is that the type check assumes that `force` always produces the same answer, but the existence of `delay/name` means that isn't true. I'll look into fixing that.

**Comment by samth** [comment]
Ok, I have a different plan. We can change the contract generated here to reject promises created with `delay/name`. That way, promises continue to typecheck they way they have in the past.

**Comment by samth** [comment]
Fixed by 5cd5f772d.
